src/pulsar_client_manager.erl

```diff
@@ -0,0 +1,381 @@
+    gen_server:call(ClientId, #get_workers{}, Timeout).
+
+%%--------------------------------------------------------------------
+%% `gen_server' API
+%%--------------------------------------------------------------------
+
+-spec init(map()) -> {ok, state()}.
+init(Args) ->
+    #{
+      client_id := ClientId,
+      seed_urls := SeedURLs,
+      initial_opts := Opts
+    } = Args,
+    process_flag(trap_exit, true),
+    State0 = #{
+        ?client_id => ClientId,
+        ?initial_opts => Opts,
+        ?seed_urls => SeedURLs,
+        ?workers => #{}
+    },
+    case spawn_any_and_wait_connected(State0) of
+        {{ok, _}, State} ->
+            {ok, State};
+        {{error, _} = Error, _State} ->
+            {stop, Error}
+    end.
+
+handle_call(#get_workers{}, _From, State) ->
+    #{?workers := Workers0} = State,
+    Workers = maps:filter(fun(_K, V) -> is_pid(V) end, Workers0),
+    {reply, Workers, State};
+handle_call(#get_status{}, _From, State0) ->
+    {Reply, State} = handle_get_status(State0),
+    {reply, Reply, State};
+handle_call(#get_alive_pulsar_url{}, _From, State0) ->
+    {Reply, State} = handle_get_alive_pulsar_url(State0),
+    {reply, Reply, State};
+handle_call(#get_topic_metadata{topic = Topic}, _From, State0) ->
+    {Reply, State} = handle_get_topic_metadata(State0, Topic),
+    {reply, Reply, State};
+handle_call(#lookup_topic{deadline = Deadline, partition_topic = PartitionTopic}, From, State0) ->
+    {Reply, State} = handle_lookup_topic_async(State0, PartitionTopic),
+    maybe
+        true ?= Deadline == infinity orelse now_ts() < Deadline,
+        gen_server:reply(From, Reply)
+    end,
+    {noreply, State};
+handle_call(_Req, _From, State) ->
+    {reply, {error, unknown_call}, State, hibernate}.
+
+handle_cast(#lookup_topic_async{from = From, partition_topic = PartitionTopic}, State0) ->
+    {Reply, State} = handle_lookup_topic_async(State0, PartitionTopic),
+    gen_server:reply(From, Reply),
+    {noreply, State};
+handle_cast(_Req, State) ->
+    {noreply, State, hibernate}.
+
+handle_info({'EXIT', Pid, Reason}, #{?workers := Workers} = State0) when
+    is_map_key(Pid, Workers)
+->
+    State = handle_worker_down(State0, Pid, Reason),
+    {noreply, State};
+handle_info(_Info, State) ->
+    log_error("received unknown message: ~p", [_Info]),
+    {noreply, State, hibernate}.
+
+terminate(_Reason, _State) ->
+    ok.
+
+%%--------------------------------------------------------------------
+%% Internal fns
+%%--------------------------------------------------------------------
+
+handle_get_status(State0) ->
+    case alive_workers(State0) of
+        [] ->
+            case spawn_any_and_wait_connected(State0) of
+                {{ok, _}, State} ->
+                    {true, State};
+                {{error, _}, State} ->
+                    {false, State}
+            end;
+        [_ | _] ->
+            %% Should we call their `get_status' to ensure they received pongs?
```

Comment:
yes, since there is no timer to tick the longtime-no-pong check in the client process.

or, if we add a timer there, then there is no need to call.